### PR TITLE
repair: Add ranges_parallelism option

### LINKED ISSUE #4847: repair: Add ranges_parallelism option
In commit 131acc09cc1d568f994059736c8500adf5d05d88 (repair: Adjust
parallelism according to memory size), we choose the ranges to repair in
parallel according to the memory size automatically.

However, this automatic number might not the optimal one. For example,
user wants repair to have minimal impact on the user cql reads and
writes, or user wants to repair at full speed allowing more cpu and
memory resources.

We can add an option that controls number of ranges that we repair 
in parallel, so that the advanced user can control the parallelism.

### PR DESCRIPTION
**Pull request by asias**
This patch adds the ranges_parallelism option to repair restful API.

Users can use this option to optionally specify the number of ranges to repair in parallel per repair job to a smaller number than the Scylla core calculated default max_repair_ranges_in_parallel.

Scylla manager can also use this option to provide more ranges (>N) in a single repair job but only repairing N ranges_parallelism in parallel, instead of providing N ranges in a repair job.

To make it safer, unlike the PR #4848, this patch does not allow user to exceed the max_repair_ranges_in_parallel.

Fixes #4847

### CODE REVIEW

**Comment by mykaul** [comment]
Do we have any tests for this option?

**Comment by denesb** [comment]
> Fixes #4847

We now have two fixes for #4847, this and #4848. Which one is the "right" one?

**Review by denesb — APPROVED** [review]
The change makes sense to me, my only problem is that there is too much stuff happening in this single commit: coroutinization, re-indent and logical changes on top. Not the end of the world.

File: repair/repair.cc [hunk 1]
```diff
@@ -949,33 +959,34 @@ future<> repair::shard_repair_task_impl::do_repair_ranges() {
...
+                nr_ranges_finished++;
+            }
+            rlogger.debug("repair[{}]: node ops progress bootstrap={}, replace={}, rebuild={}, decommission={}, removenode={}, repair={}",
+                global_repair_id.uuid(),
+                rs.get_metrics().bootstrap_finished_percentage(),
+                rs.get_metrics().replace_finished_percentage(),
+                rs.get_metrics().rebuild_finished_percentage(),
+                rs.get_metrics().decommission_finished_percentage(),
+                rs.get_metrics().removenode_finished_percentage(),
+                rs.get_metrics().repair_finished_percentage());
```
[the remark this answers is not in the document]
Comment (asias) [reply repair/repair.cc:989]: I usually use `git log -p --ignore-all-space` to  ignore space change noises.  E.g., the change is very clean with it. 

```
@@ -949,9 +959,12 @@ future<> repair::shard_repair_task_impl::do_repair_ranges() {
         // repair all the ranges in limited parallelism
         rlogger.info("repair[{}]: Started to repair {} out of {} tables in keyspace={}, table={}, table_id={}, repair_reason={}",
                 global_repair_id.uuid(), idx + 1, table_ids.size(), _status.keyspace, table_name, table_id, _reason);
-        co_await coroutine::parallel_for_each(ranges, [this, table_id] (auto&& range) {
-            return with_semaphore(rs.get_repair_module().range_parallelism_semaphore(), 1, [this, &range, table_id] {
-                return repair_range(range, table_id).then([this] {
+        co_await coroutine::parallel_for_each(ranges, [this, table_id] (auto&& range) -> future<> {
+            // Get the system range parallelism
+            auto permit = co_await seastar::get_units(rs.get_repair_module().range_parallelism_semaphore(), 1);
+            // Get the range parallelism specified by user
+            auto user_permit = _user_ranges_parallelism ? co_await seastar::get_units(*_user_ranges_parallelism, 1) : semaphore_units<>();
+            co_await repair_range(range, table_id);
             if (_reason == streaming::stream_reason::bootstrap) {
                 rs.get_metrics().bootstrap_finished_ranges++;
             } else if (_reason == streaming::stream_reason::replace) {
@@ -975,8 +988,6 @@ future<> repair::shard_repair_task_impl::do_repair_ranges() {
                 rs.get_metrics().removenode_finished_percentage(),
                 rs.get_metrics().repair_finished_percentage());
         });
-            });
-        });
 
         if (_reason != streaming::stream_reason::repair) {
             try {
@@ -1129,7 +1140,8 @@ future<int> repair_service::do_repair_start(sstring keyspace, std::unordered_map
```

**Comment by asias** [comment]
> > Fixes #4847
> 
> We now have two fixes for #4847, this and #4848. Which one is the "right" one?

This one replaces #4848. I closed the old one.

File: repair/repair.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (asias) [reply repair/repair.cc:989]: In v2, I split the coroutine change to a separate patch.

**Comment by asias** [comment]
> Do we have any tests for this option?

The test will follow.  I am still thinking about how to test it more efficiently.

**Comment by asias** [comment]
@denesb  Ping for merge.

**Comment by vladzcloudius** [comment]
@asias, just to confirm - this configurable effect is applied to RBNO as well, right?

**Comment by asias** [comment]
> @asias, just to confirm - this configurable effect is applied to RBNO as well, right?

This is an option specified by the restful api to start user requested repair. So it does not applies to RBNO.

**Comment by tzach** [comment]
@asias can this change effect the way Scylla Manager use the repair?
@karol-kokoszka FYI

**Comment by asias** [comment]
> @asias can this change effect the way Scylla Manager use the repair? @karol-kokoszka FYI

Yes, as I mentioned in other threads. SM can use this option to provide X ranges in each api request but only start Y (X > Y) ranges in parallel.

**Pull request merged**